### ISSUE TITLE
PostBack.ConcurrencyMode property

### ISSUE THREAD

**Issue by tomasherceg** [riganti/dotvvm#283]
We need to be able to specify concurrency mode for a situations where multiple postbacks occur simultaneously.

We should add an attached property `PostBack.ConcurrencyMode` with the following values:

* `LastWins`: When the client receives a response for a particular postback while another postback has been already invoked, the response is ignored. It's the "last postback wins" strategy.
* `Deny`: When a postback is still running, you cannot invoke another one - it will be dropped. The "only one postback" strategy. 
* `Queue`: When a postback is still running and you invoke another one, it will wait in the queue and invoke immediately after the response from the current postback is received. If the current postback changes something in the viewmodel, the new postback must use the updated viewmodel.

If you set this property on any element, its value will be propagated to the children.

**Comment by martindybal** [comment]
There should be also property in config for default behavior

**Comment by tomasherceg** [comment]
Since the value of this property should propagate to the child elements, you can set the app default e.g. on the `body` element in the master page.

**Comment by exyi** [comment]
I have just discussed this feature with @tomasherceg, so here are the results:

The concurrency mode will only affect how is the postback launched, that means:
* `Queue` - launch the postback if no postback is running (including `Deny` and `None`) or add it to the postback queue.
* `Deny` - launch the postback if no postback is running or drop it
* `None` - just launch the postback, equivalent to the current behavior

When the response from the server is received, it's always dropped if newer postback has already launched. That means that during the `Queue` postback runtime, `None` can launch and cause rejection of the previous one:
```
|Queue Launched                                  |Done and result is dropped
|>-----------------------------------------------|
        
         |None Launched                                                        |Commited
         |>--------------------------------------------------------------------|
```

The postbacks can be isolated by specifying a "postback queue" name - postback in different queues will be allowed to run in parallel and both will be committed. That could lead to pretty strange behavior if both would change the same part of the view model, but that the programmer's responsibility to consider.

Static commands results will not be dropped when executed in parallel, but the concurrency mode will apply to them as well.

**Comment by djanosik** [comment]
How will it work when I have a command where I always want to use `Queue`? Any way to ignore user settings?

**Comment by exyi** [comment]
@djanosik You can't specify the concurrency mode on the invoked method, it has to be on the control that contains the command. So if you have a custom control that contains the command, you can overwrite the concurrency mode from the parent, I don't think it will be easily possible to implement this option for a specific method.

**Comment by djanosik** [comment]
Ok. Can we also implement some rate limit for postbacks? Eg. send the postback after not being sent for 200ms? It would be up to control author to set rate limit for command or not.

**Comment by exyi** [comment]
I think that you can use a custom Postback Handler for rate limiting.
The concurrency mode is basically also a postback handler, so the rate
limiting should not be a problem to implement.

**Comment by tomasherceg** [comment]
Closing, let'S add rate limiting as a separate issue.
